dg.go

```diff
@@ -244,16 +244,17 @@ func (d *directedGraph[NodeType]) HasReadyNodes() bool {
 }
 
 func (d *directedGraph[NodeType]) PopReadyNodes() map[string]ResolutionStatus {
-	d.lock.Lock()
-	// Transfer the map to a local variable to minimize time locked, and reset the graph's value.
-	readyMap := d.readyForProcessing
-	d.readyForProcessing = make(map[string]*node[NodeType])
-	d.lock.Unlock()
-
 	result := make(map[string]ResolutionStatus)
-	for _, node := range readyMap {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	for _, node := range d.readyForProcessing {
+		// There are some conditions that may result in the status being modified,
+		// so this needs to be done under lock.
+		// For example, a ready waiting node being marked Resolved or Unresolvable by
+		// a user that retrieves the node by ID.
```

Comment:
This code isn't changing anything, just copying node `status`. Either (1) it's important that the set of status values returned are a consistent snapshot of the DAG, or (2) this is just another spurious Go race detection workaround. Either way this comment doesn't seem particularly helpful.

Reply:
What this lock does is two things.

First, it ensures that the returned status values are from the point in time when the ready list was popped.  They might have been changed after that, but the caller will be none the wiser.  (Granted, this is not materially different from doing the copy without the lock, but....)

Second, it defeats Go's race condition detection (which, really, isn't an actual race detector as much as a potential race detector, which returns a false positive, here).

So, I think using the lock here (to avoid the false positive) is good.  The comment, however, I don't think is helpful.  As [I said above](https://github.com/arcalot/go-dgraph/pull/17#discussion_r1657861582), I think we should just delete it.

Reply:
The comment is also useful to me. If in the future I wonder why the lock is done that way, it's useful context. I wouldn't want to remove it for my own long-term sake.